
The Findlay Function Scores tile displays functional assay results for variants in functionally  critical domains of BRCA1, obtained using saturation genome editing. The data is provided freely for nonprofit use. Commercial entities should contact the University of Washington for licensing agreements.

Nearly all possible single nucleotide variants (SNVs) within specific regions of BRCA1 were introduced using saturation genome editing (SGE). SGE utilizes CRISPR-Cas9 technology to introduce SNVs in near-haploid human cells (HAP1). Variant impact on cell survival and mRNA expression levels was then assessed. The function score measures the degree of cell death observed when the variant was introduced.

The function scores for all variants in the experiment are displayed in a histogram. This histogram, based on Figure 2g, represents all 3,893 variants assayed (1). Each variant scored in an SGE experiment has its own function score: this score is indicated by an arrow under the appropriate histogram bar. The X-axis of the histogram contains function scores, while the Y-axis of the histogram denotes how many total variants assayed fall within each score range.

These function scores are bimodally distributed. A variant with function score will be found in lower range is reported to be associated with loss of function, while a score in the higher range is associated with no functional impact. Variants with scores in the “intermediate” range were not definitively assessed in this assay. These thresholds were developed by the authors of this assay method, and adapted for display on BRCA Exchange.

For more information, please refer to the original publication (1).

1. [Findlay et al. 2018. PMID 30209399.](https://www.ncbi.nlm.nih.gov/pubmed/18992264)

Publications that report on the accuracy of this assay:
- [Findlay et al. 2018. PMID 30209399.](https://www.ncbi.nlm.nih.gov/pubmed/18992264)
