<html><head><meta content="text/html; charset=UTF-8" http-equiv="content-type"><style type="text/css">ul.lst-kix_exint55jajpv-7{list-style-type:none}ul.lst-kix_exint55jajpv-6{list-style-type:none}ul.lst-kix_exint55jajpv-5{list-style-type:none}ul.lst-kix_exint55jajpv-4{list-style-type:none}ul.lst-kix_exint55jajpv-8{list-style-type:none}ul.lst-kix_isewf3psavwj-8{list-style-type:none}.lst-kix_exint55jajpv-4>li:before{content:"\0025cb   "}.lst-kix_exint55jajpv-5>li:before{content:"\0025a0   "}ul.lst-kix_isewf3psavwj-7{list-style-type:none}ul.lst-kix_isewf3psavwj-6{list-style-type:none}ul.lst-kix_isewf3psavwj-5{list-style-type:none}.lst-kix_exint55jajpv-6>li:before{content:"\0025cf   "}.lst-kix_exint55jajpv-8>li:before{content:"\0025a0   "}ul.lst-kix_isewf3psavwj-0{list-style-type:none}ul.lst-kix_isewf3psavwj-4{list-style-type:none}ul.lst-kix_isewf3psavwj-3{list-style-type:none}.lst-kix_exint55jajpv-7>li:before{content:"\0025cb   "}ul.lst-kix_isewf3psavwj-2{list-style-type:none}ul.lst-kix_isewf3psavwj-1{list-style-type:none}.lst-kix_isewf3psavwj-7>li:before{content:"\0025cb   "}.lst-kix_ikaq0minelf9-8>li:before{content:"\0025a0   "}.lst-kix_isewf3psavwj-8>li:before{content:"\0025a0   "}.lst-kix_isewf3psavwj-5>li:before{content:"\0025a0   "}.lst-kix_isewf3psavwj-6>li:before{content:"\0025cf   "}.lst-kix_ikaq0minelf9-1>li:before{content:"\0025cb   "}.lst-kix_ikaq0minelf9-0>li:before{content:"  "}ul.lst-kix_ikaq0minelf9-1{list-style-type:none}ul.lst-kix_ikaq0minelf9-2{list-style-type:none}ul.lst-kix_ikaq0minelf9-0{list-style-type:none}.lst-kix_exint55jajpv-3>li:before{content:"\0025cf   "}li.li-bullet-0:before{margin-left:-18pt;white-space:nowrap;display:inline-block;min-width:18pt}.lst-kix_ikaq0minelf9-2>li:before{content:"\0025a0   "}.lst-kix_ikaq0minelf9-3>li:before{content:"\0025cf   "}.lst-kix_isewf3psavwj-1>li:before{content:"\0025cb   "}.lst-kix_ikaq0minelf9-5>li:before{content:"\0025a0   "}.lst-kix_exint55jajpv-2>li:before{content:"\0025a0   "}.lst-kix_isewf3psavwj-0>li:before{content:"\0025cf   "}.lst-kix_ikaq0minelf9-4>li:before{content:"\0025cb   "}.lst-kix_isewf3psavwj-4>li:before{content:"\0025cb   "}ul.lst-kix_ikaq0minelf9-5{list-style-type:none}.lst-kix_ikaq0minelf9-7>li:before{content:"\0025cb   "}.lst-kix_exint55jajpv-0>li:before{content:"\0025cf   "}.lst-kix_exint55jajpv-1>li:before{content:"\0025cb   "}ul.lst-kix_ikaq0minelf9-6{list-style-type:none}ul.lst-kix_ikaq0minelf9-3{list-style-type:none}.lst-kix_isewf3psavwj-3>li:before{content:"\0025cf   "}ul.lst-kix_ikaq0minelf9-4{list-style-type:none}ul.lst-kix_exint55jajpv-3{list-style-type:none}ul.lst-kix_exint55jajpv-2{list-style-type:none}ul.lst-kix_exint55jajpv-1{list-style-type:none}.lst-kix_ikaq0minelf9-6>li:before{content:"\0025cf   "}ul.lst-kix_ikaq0minelf9-7{list-style-type:none}ul.lst-kix_exint55jajpv-0{list-style-type:none}.lst-kix_isewf3psavwj-2>li:before{content:"\0025a0   "}ul.lst-kix_ikaq0minelf9-8{list-style-type:none}ol{margin:0;padding:0}table td,table th{padding:0}.c5{margin-left:72pt;padding-top:0pt;padding-left:0pt;padding-bottom:0pt;line-height:1.15;orphans:2;widows:2;text-align:left}.c7{padding-top:0pt;padding-bottom:0pt;line-height:1.15;orphans:2;widows:2;text-align:left;height:11pt}.c1{color:#000000;font-weight:400;text-decoration:none;vertical-align:baseline;font-size:20pt;font-family:"Arial";font-style:normal}.c11{color:#000000;font-weight:700;text-decoration:none;vertical-align:baseline;font-size:11pt;font-family:"Arial";font-style:normal}.c0{color:#000000;font-weight:400;text-decoration:none;vertical-align:baseline;font-size:11pt;font-family:"Arial";font-style:normal}.c9{padding-top:20pt;padding-bottom:6pt;line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}.c13{padding-top:0pt;padding-bottom:0pt;line-height:1.15;orphans:2;widows:2;text-align:left}.c15{text-decoration-skip-ink:none;-webkit-text-decoration-skip:none;color:#1155cc;text-decoration:underline}.c23{background-color:#ffffff;max-width:468pt;padding:72pt 72pt 72pt 72pt}.c4{color:inherit;text-decoration:inherit}.c21{margin-left:36pt;padding-left:0pt}.c2{padding:0;margin:0}.c10{color:#646464;font-weight:700}.c14{color:#00b455;font-weight:700}.c8{color:#1e0c2b;font-weight:700}.c16{color:#0d55bf;font-weight:700}.c3{color:#239f86;font-weight:700}.c12{color:#025242;font-weight:700}.c27{color:#583a6e}.c6{font-style:italic}.c26{color:#8a6f9e}.c19{color:#a8e0d5}.c17{font-weight:700}.c25{color:#7cbfb2}.c29{color:#9c0506}.c24{margin-left:72pt}.c18{color:#7d7d7d}.c22{color:#434343}.c28{color:#212121}.c20{color:#ccbad9}.title{padding-top:0pt;color:#000000;font-size:26pt;padding-bottom:3pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}.subtitle{padding-top:0pt;color:#666666;font-size:15pt;padding-bottom:16pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}li{color:#000000;font-size:11pt;font-family:"Arial"}p{margin:0;color:#000000;font-size:11pt;font-family:"Arial"}h1{padding-top:20pt;color:#000000;font-size:20pt;padding-bottom:6pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h2{padding-top:18pt;color:#000000;font-size:16pt;padding-bottom:6pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h3{padding-top:16pt;color:#434343;font-size:14pt;padding-bottom:4pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h4{padding-top:14pt;color:#666666;font-size:12pt;padding-bottom:4pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h5{padding-top:12pt;color:#666666;font-size:11pt;padding-bottom:4pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h6{padding-top:12pt;color:#666666;font-size:11pt;padding-bottom:4pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;font-style:italic;orphans:2;widows:2;text-align:left}</style></head><body class="c23 doc-content"><h1 class="c9" id="h.986c7pzdvgr0"><span class="c1">Description</span></h1><p class="c7"><span class="c0"></span></p><p class="c13"><span>The </span><span class="c15"><a class="c4" href="https://www.google.com/url?q=https://brcaexchange.org/&amp;sa=D&amp;source=editors&amp;ust=1723755805325664&amp;usg=AOvVaw2Df8zKvu4E_2ausMbklDwQ">BRCA Exchange</a></span><span>&nbsp;web portal was developed to share known, expert-curated knowledge on the germline variants in </span><span class="c6">BRCA1</span><span>&nbsp;and </span><span class="c6">BRCA2</span><span class="c0">, and how these variants may affect an individual&rsquo;s risk of heritable breast, ovarian, pancreatic and other cancers (HBOC). &nbsp;This set of variants is obtained by integrating data from genomic variation repositories including ClinVar, LOVD and gnomAD, and annotating these variants according to methods selected and developed in close collaboration with the ENIGMA Consortium, the international expert consortium on research on HBOC variants. &nbsp; This track hub presents the variants from BRCA Exchange in the context of the human genome, offering further details on each variant in the variant details pages.</span></p><p class="c7"><span class="c0"></span></p><p class="c13"><span class="c0">The available tracks are:</span></p><ul class="c2 lst-kix_isewf3psavwj-0 start"><li class="c13 c21 li-bullet-0"><span class="c0">BRCA Variants - Shows single nucleotide variants (SNVs) and small insertion/deletion variants of &lt; 50 nucleotides (indels) </span></li><li class="c13 c21 li-bullet-0"><span class="c0">BRCA Structural Variants - Shows variants &gt;=50 nucleotides</span></li><li class="c13 c21 li-bullet-0"><span class="c0">ACMG Popfreq Variants - shows the provisionally-assigned ACMG evidence code for the Population Frequency group, for variants of &lt; 50 nucleotides</span></li><li class="c13 c21 li-bullet-0"><span class="c0">ACMG Popfreq SVs - shows the provisionally-assigned ACMG evidence code for the Population Frequency group, for variants of &gt;= 50 nucleotides</span></li></ul><h1 class="c9" id="h.kpklhyqaux3"><span class="c1">Genomes</span></h1><p class="c7"><span class="c0"></span></p><p class="c13"><span>We provide data on </span><span class="c6">BRCA </span><span class="c0">variants for the GRCh37 and GRCh38 genomes.</span></p><h1 class="c9" id="h.mjc1249x4n1u"><span class="c1">Display Conventions and Configuration</span></h1><p class="c13"><span class="c0">Each variant from BRCA Exchange is represented as one item in this track, with boxes indicating the start and end &nbsp;genomic coordinates of each variant. &nbsp;Clicking on a variant will take one to the variant&rsquo;s track details page, with additional variant data. &nbsp;Further details on these data are available through the BRCA Exchange Help pages.</span></p><p class="c7"><span class="c11"></span></p><ul class="c2 lst-kix_exint55jajpv-0 start"><li class="c13 c21 li-bullet-0"><span class="c17">BRCA Variants</span><span>&nbsp;and </span><span class="c17">BRCA Structural Variants</span></li></ul><ul class="c2 lst-kix_exint55jajpv-1 start"><li class="c5 li-bullet-0"><span>Items in </span><span class="c17 c29">red</span><span class="c0">&nbsp;represent variants classified by the ENIGMA VCEP as Pathogenic or Likely Pathogenic. &nbsp;</span></li><li class="c5 li-bullet-0"><span>Items in </span><span class="c14">green</span><span class="c0">&nbsp;represent variants classified by the ENIGMA VCEP as Benign or Likely Benign. &nbsp;</span></li><li class="c5 li-bullet-0"><span>Items in </span><span class="c10">gray</span><span class="c18">&nbsp;</span><span class="c22">represent variants which have</span><span class="c0">&nbsp;not yet been reviewed by the ENIGMA VCEP.</span></li></ul><ul class="c2 lst-kix_exint55jajpv-0"><li class="c13 c21 li-bullet-0"><span class="c17">VCEP Curation Evidence: </span><span>Items that convey evidence towards a pathogenic interpretation are colored purple. &nbsp;Items that convey evidence towards a benign interpretation are colored cyan. &nbsp;In both cases, the shade of the color indicates the strength of evidence, with darker colors representing stronger evidence. &nbsp;</span><span class="c11">This includes Population Frequency.</span></li></ul><ul class="c2 lst-kix_exint55jajpv-1 start"><li class="c5 li-bullet-0"><span class="c8">Dark purple</span><span>&nbsp;indicates Pathogenic Very Strong evidence, while </span><span class="c17 c27">deep purple</span><span>&nbsp;indicates Pathogenic Strong, </span><span class="c17 c26">light purple</span><span>&nbsp;indicates Pathogenic Moderate, and </span><span class="c17 c20">very light purple</span><span class="c0">&nbsp;indicates Pathogenic Supporting.</span></li><li class="c5 li-bullet-0"><span class="c12">Dark cyan</span><span>&nbsp;indicates Benign Very Strong or Standalone evidence, while </span><span class="c3">deep cyan</span><span>&nbsp;indicates Benign Strong, </span><span class="c17 c25">light cyan</span><span>&nbsp;indicates Benign Moderate, and </span><span class="c17 c19">very light cyan</span><span class="c0">&nbsp;indicates Benign Supporting.</span></li><li class="c5 li-bullet-0"><span class="c16">Deep blue</span><span class="c0">&nbsp;indicates an evidence item with no code assigned, either because it failed to meet the evidence thresholds or because of a technical issue.</span></li></ul><p class="c7 c24"><span class="c0"></span></p><h1 class="c9" id="h.7di0evggxb85"><span class="c1">Methods</span></h1><p class="c7"><span class="c0"></span></p><p class="c13"><span>The BRCA Exchange variant set is generated monthly, by combining lists of known variants from ClinVar, the gnomAD 2.1 and 3.1 non-cancer subsets, LOVD, The BRCA ExUV resource, and sets of functional assays curated by the ENIGMA Consortium. &nbsp;Data from these resources are integrated by comparing the GRCh38 genomic coordinates and reference and alternative alleles, merging equivalent variants. &nbsp;These variants are then annotated by a pipeline that computes </span><span class="c6">in silico</span><span class="c0">&nbsp;prediction scores from methods including from BayesDel, AlignGVGD and SpliceAI. &nbsp;For more information, see the Help documentation on BRCA Exchange.</span></p><p class="c7"><span class="c0"></span></p><h1 class="c9" id="h.lh4mydunil7l"><span class="c1">Data Availability</span></h1><p class="c7"><span class="c0"></span></p><p class="c13"><span>All data are freely available for download via </span><span class="c15"><a class="c4" href="https://www.google.com/url?q=https://brcaexchange.org/releases&amp;sa=D&amp;source=editors&amp;ust=1723755805329022&amp;usg=AOvVaw2tU-4ClPhtaLY2oPlUrF65">https://brcaexchange.org/releases</a></span><span>,</span><span class="c0">&nbsp; &nbsp;</span></p><h1 class="c9" id="h.xukp72k0k5sd"><span class="c1">Credits</span></h1><p class="c7"><span class="c0"></span></p><p class="c13"><span>BRCA Exchange was developed by the BRCA Challenge Consortium, which was launched by the GA4GH to develop responsible approaches for sharing knowledge on genetic variation and disease. &nbsp;BRCA Exchange was implemented by the labs of Melissa Cline at UC Santa Cruz, Benedict &nbsp;Paten at UC Santa Cruz and Gunnar R&auml;tsch at ETH Zurich, in close consultation with Amanda Spurdle at QIMR Berghofer. &nbsp;The Cline and Spurdle labs continue the on-going development of the resource. &nbsp;For questions, please contact </span><span class="c15"><a class="c4" href="mailto:brcaexchange@gmail.com">brcaexchange@gmail.com</a></span><span class="c0">.</span></p><p class="c7"><span class="c0"></span></p><p class="c13"><span>ENIGMA Consortium: </span><span class="c15"><a class="c4" href="https://www.google.com/url?q=https://enigmaconsortium.org/&amp;sa=D&amp;source=editors&amp;ust=1723755805330048&amp;usg=AOvVaw1IIbs41Z7NAgDfqemcCyzE">https://enigmaconsortium.org/</a></span></p><p class="c7"><span class="c0"></span></p><p class="c13"><span>ENIGMA BRCA1 and BRCA2 Variant Curation Expert Panel: </span><span class="c15"><a class="c4" href="https://www.google.com/url?q=https://clinicalgenome.org/affiliation/50087/&amp;sa=D&amp;source=editors&amp;ust=1723755805330383&amp;usg=AOvVaw1J5JtVnr0-mK_ATGFGh3_0">https://clinicalgenome.org/affiliation/50087/</a></span></p><p class="c7"><span class="c0"></span></p><p class="c7"><span class="c0"></span></p><p class="c7"><span class="c0"></span></p><h1 class="c9" id="h.vs6575nqzuat"><span class="c1">References</span></h1><p class="c7"><span class="c0"></span></p><p class="c13"><span>Cline MS, Liao RG, Parsons MT, Paten B </span><span class="c6">et al</span><span>. </span><span class="c15"><a class="c4" href="https://www.google.com/url?q=https://pubmed.ncbi.nlm.nih.gov/30586411/&amp;sa=D&amp;source=editors&amp;ust=1723755805330961&amp;usg=AOvVaw3uYTBmfDAcsA3tmj_Y3aAJ">BRCA Challenge: BRCA Exchange as a global resource for variants in BRCA1 and BRCA2</a></span><span>, </span><span class="c6">PLoS Genet</span><span>. 2018 Dec 26;14(12):e1007752. doi: </span><span class="c15"><a class="c4" href="https://www.google.com/url?q=https://doi.org/10.1371/journal.pgen.1007752&amp;sa=D&amp;source=editors&amp;ust=1723755805331227&amp;usg=AOvVaw3btdvr6OGxCcweMJjo9PXf">10.1371/journal.pgen.1007752</a></span><span>. PMID: </span><span class="c15"><a class="c4" href="https://www.google.com/url?q=https://pubmed.ncbi.nlm.nih.gov/30586411/&amp;sa=D&amp;source=editors&amp;ust=1723755805331456&amp;usg=AOvVaw1ZRrsG1jKgerBA87iRHOW5">30586411</a></span><span>; &nbsp;</span><span class="c28">PMCID:</span><span class="c15"><a class="c4" href="https://www.google.com/url?q=http://www.ncbi.nlm.nih.gov/pmc/articles/pmc6324924/&amp;sa=D&amp;source=editors&amp;ust=1723755805331702&amp;usg=AOvVaw2jqxkJVkzEqoC4P3H-I4EC">&nbsp;PMC6324924</a></span></p><p class="c7"><span class="c0"></span></p><p class="c13"><span>Parsons MT, de la Hoya M, Richardson ME, Tudini E</span><span class="c6">&nbsp;et al.</span><span>&nbsp;</span><span class="c15"><a class="c4" href="https://www.google.com/url?q=https://www.cell.com/ajhg/abstract/S0002-9297(24)00257-X&amp;sa=D&amp;source=editors&amp;ust=1723755805332115&amp;usg=AOvVaw1yqCd7f0yyvu3LX-wWSOkQ">Evidence-based recommendations for gene-specific ACMG/AMP variant classification from the ClinGen ENIGMA BRCA1 and BRCA2 Variant Curation Expert Panel</a></span><span>, </span><span class="c6">ASHG</span><span>&nbsp;2024 Aug 13. doi:</span><span class="c15"><a class="c4" href="https://www.google.com/url?q=https://doi.org/10.1016/j.ajhg.2024.07.013&amp;sa=D&amp;source=editors&amp;ust=1723755805332350&amp;usg=AOvVaw2U6u2d7nzsxwh9CJOzhXlw">10.1016/j.ajhg.2024.07.013</a></span><span class="c0">.</span></p><p class="c7"><span class="c0"></span></p><p class="c7"><span class="c0"></span></p><p class="c7"><span class="c0"></span></p><p class="c7"><span class="c0"></span></p><p class="c7"><span class="c0"></span></p><p class="c7"><span class="c0"></span></p><p class="c7"><span class="c0"></span></p></body></html>