<html><head><meta content="text/html; charset=UTF-8" http-equiv="content-type"><style type="text/css">ul.lst-kix_ikaq0minelf9-1{list-style-type:none}ul.lst-kix_ikaq0minelf9-2{list-style-type:none}.lst-kix_ikaq0minelf9-8>li:before{content:"\0025a0   "}ul.lst-kix_ikaq0minelf9-0{list-style-type:none}.lst-kix_ikaq0minelf9-2>li:before{content:"\0025a0   "}.lst-kix_ikaq0minelf9-3>li:before{content:"\0025cf   "}.lst-kix_ikaq0minelf9-1>li:before{content:"\0025cb   "}.lst-kix_ikaq0minelf9-5>li:before{content:"\0025a0   "}.lst-kix_ikaq0minelf9-0>li:before{content:"  "}.lst-kix_ikaq0minelf9-4>li:before{content:"\0025cb   "}ul.lst-kix_ikaq0minelf9-5{list-style-type:none}.lst-kix_ikaq0minelf9-7>li:before{content:"\0025cb   "}ul.lst-kix_ikaq0minelf9-6{list-style-type:none}ul.lst-kix_ikaq0minelf9-3{list-style-type:none}ul.lst-kix_ikaq0minelf9-4{list-style-type:none}.lst-kix_ikaq0minelf9-6>li:before{content:"\0025cf   "}ul.lst-kix_ikaq0minelf9-7{list-style-type:none}ul.lst-kix_ikaq0minelf9-8{list-style-type:none}ol{margin:0;padding:0}table td,table th{padding:0}.c7{padding-top:20pt;padding-bottom:6pt;line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}.c2{color:#000000;font-weight:400;text-decoration:none;vertical-align:baseline;font-size:20pt;font-family:"Arial";font-style:normal}.c4{color:#000000;font-weight:400;text-decoration:none;vertical-align:baseline;font-size:11pt;font-family:"Arial";font-style:normal}.c0{padding-top:0pt;padding-bottom:0pt;line-height:1.15;orphans:2;widows:2;text-align:left}.c5{text-decoration-skip-ink:none;-webkit-text-decoration-skip:none;color:#1155cc;text-decoration:underline}.c9{background-color:#ffffff;max-width:468pt;padding:72pt 72pt 72pt 72pt}.c6{color:inherit;text-decoration:inherit}.c1{font-style:italic}.c3{height:11pt}.c8{color:#212121}.title{padding-top:0pt;color:#000000;font-size:26pt;padding-bottom:3pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}.subtitle{padding-top:0pt;color:#666666;font-size:15pt;padding-bottom:16pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}li{color:#000000;font-size:11pt;font-family:"Arial"}p{margin:0;color:#000000;font-size:11pt;font-family:"Arial"}h1{padding-top:20pt;color:#000000;font-size:20pt;padding-bottom:6pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h2{padding-top:18pt;color:#000000;font-size:16pt;padding-bottom:6pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h3{padding-top:16pt;color:#434343;font-size:14pt;padding-bottom:4pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h4{padding-top:14pt;color:#666666;font-size:12pt;padding-bottom:4pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h5{padding-top:12pt;color:#666666;font-size:11pt;padding-bottom:4pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h6{padding-top:12pt;color:#666666;font-size:11pt;padding-bottom:4pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;font-style:italic;orphans:2;widows:2;text-align:left}</style></head><body class="c9 doc-content"><h1 class="c7" id="h.986c7pzdvgr0"><span class="c2">Description</span></h1><p class="c0 c3"><span class="c4"></span></p><p class="c0"><span>The </span><span class="c5"><a class="c6" href="https://www.google.com/url?q=https://brcaexchange.org/&amp;sa=D&amp;source=editors&amp;ust=1709092395400354&amp;usg=AOvVaw2115XuabWg5jqum6L4iOm6">BRCA Exchange</a></span><span>&nbsp;web portal was developed to share known, expert-curated knowledge on the germline variants in </span><span class="c1">BRCA1</span><span>&nbsp;and </span><span class="c1">BRCA2</span><span class="c4">, and how these variants may affect an individual&rsquo;s risk of heritable breast, ovarian, pancreatic and other cancers (HBOC). &nbsp;This set of variants is obtained by integrating data from genomic variation repositories including ClinVar, LOVD and gnomAD, and annotating these variants according to methods selected and developed in close collaboration with the ENIGMA Consortium, the international expert consortium on research on HBOC variants. &nbsp; This track hub presents the variants from BRCA Exchange in the context of the human genome, offering further details on each variant in the variant details pages.</span></p><h1 class="c7" id="h.kpklhyqaux3"><span class="c2">Genomes</span></h1><p class="c0 c3"><span class="c4"></span></p><p class="c0"><span>We provide data on </span><span class="c1">BRCA </span><span class="c4">variants for the GRCh37 and GRCh38 genomes.</span></p><h1 class="c7" id="h.mjc1249x4n1u"><span class="c2">Display Conventions and Configuration</span></h1><p class="c0 c3"><span class="c4"></span></p><p class="c0"><span class="c4">Each variant from BRCA Exchange is represented as one item in this track, with boxes indicating the start and end &nbsp;genomic coordinates of each variant. &nbsp;Clicking on a variant will take one to the variant&rsquo;s track details page, which displays all of the available data per variant. &nbsp;Further details on these data are available through the BRCA Exchange Help pages.</span></p><p class="c0 c3"><span class="c4"></span></p><h1 class="c7" id="h.7di0evggxb85"><span class="c2">Methods</span></h1><p class="c0 c3"><span class="c4"></span></p><p class="c0"><span>The BRCA Exchange variant set is generated monthly, by combining lists of known variants from ClinVar, the gnomAD 2.1 and 3.1 non-cancer subsets, LOVD, The BRCA ExUV resource, and sets of functional assays curated by the ENIGMA Consortium. &nbsp;Data from these resources are integrated by comparing the GRCh38 genomic coordinates and reference and alternative alleles, merging equivalent variants. &nbsp;These variants are then annotated by a pipeline that computes </span><span class="c1">in silico</span><span class="c4">&nbsp;prediction scores from methods including from BayesDel, AlignGVGD and SpliceAI. &nbsp;For more information, see the Help documentation on BRCA Exchange.</span></p><p class="c0 c3"><span class="c4"></span></p><h1 class="c7" id="h.lh4mydunil7l"><span class="c2">Data Availability</span></h1><p class="c0 c3"><span class="c4"></span></p><p class="c0"><span>All data are freely available for download via </span><span class="c5"><a class="c6" href="https://www.google.com/url?q=https://brcaexchange.org/releases&amp;sa=D&amp;source=editors&amp;ust=1709092395401709&amp;usg=AOvVaw0n7sXL61M3rCjWb7wSOIJl">https://brcaexchange.org/releases</a></span><span>,</span><span class="c4">&nbsp; &nbsp;</span></p><h1 class="c7" id="h.xukp72k0k5sd"><span class="c2">Credits</span></h1><p class="c0 c3"><span class="c4"></span></p><p class="c0"><span>BRCA Exchange was developed by the BRCA Challenge Consortium, which was launched by the GA4GH to develop responsible approaches for sharing knowledge on genetic variation and disease. &nbsp;BRCA Exchange was implemented by the labs of Melissa Cline at UC Santa Cruz, Benedict &nbsp;Paten at UC Santa Cruz and Gunnar R&auml;tsch at ETH Zurich, in close consultation with Amanda Spurdle at QIMR Berghofer. &nbsp;The Cline and Spurdle labs continue the on-going development of the resource. &nbsp;For questions, please contact </span><span class="c5"><a class="c6" href="mailto:brca-exchange-contact@genomicsandhealth.org">brca-exchange-contact@genomicsandhealth.org</a></span><span class="c4">.</span></p><p class="c0 c3"><span class="c4"></span></p><h1 class="c7" id="h.vs6575nqzuat"><span class="c2">References</span></h1><p class="c0 c3"><span class="c4"></span></p><p class="c0"><span>Cline MS, Liao RG, Parsons MT, Paten B </span><span class="c1">et al</span><span>. </span><span class="c5"><a class="c6" href="https://www.google.com/url?q=https://pubmed.ncbi.nlm.nih.gov/30586411/&amp;sa=D&amp;source=editors&amp;ust=1709092395402660&amp;usg=AOvVaw1ucRQBTXG-T8OrdSmPM2vp">BRCA Challenge: BRCA Exchange as a global resource for variants in BRCA1 and BRCA2</a></span><span>, </span><span class="c1">PLoS Genet</span><span>. 2018 Dec 26;14(12):e1007752. doi: </span><span class="c5"><a class="c6" href="https://www.google.com/url?q=https://doi.org/10.1371/journal.pgen.1007752&amp;sa=D&amp;source=editors&amp;ust=1709092395402876&amp;usg=AOvVaw3ZHsVHTtYvkqJp4GO1hduh">10.1371/journal.pgen.1007752</a></span><span>. PMID: </span><span class="c5"><a class="c6" href="https://www.google.com/url?q=https://pubmed.ncbi.nlm.nih.gov/30586411/&amp;sa=D&amp;source=editors&amp;ust=1709092395403047&amp;usg=AOvVaw3oexTbNHjSZ-zDhJRk-XVr">30586411</a></span><span>; &nbsp;</span><span class="c8">PMCID:</span><span class="c5"><a class="c6" href="https://www.google.com/url?q=http://www.ncbi.nlm.nih.gov/pmc/articles/pmc6324924/&amp;sa=D&amp;source=editors&amp;ust=1709092395403236&amp;usg=AOvVaw2VO1Q-Pi5KyGkM-zJ54ev-">&nbsp;PMC6324924</a></span></p><p class="c0 c3"><span class="c4"></span></p><p class="c0 c3"><span class="c4"></span></p><p class="c0 c3"><span class="c4"></span></p><p class="c0 c3"><span class="c4"></span></p><p class="c0 c3"><span class="c4"></span></p><p class="c0 c3"><span class="c4"></span></p></body></html>